### ISSUE TITLE
Eliminating conversions to unscoped enum type

### ISSUE THREAD

**Issue by aantron** [aantron/better-enums#25]
Figured I'd open an issue for this, since it was discussed in multiple places (#22, #23).

I have another suggestion besides making "strict conversions" the default: eliminate implicit conversions to enums completely. IIRC they are only needed for `switch`, for which we could provide a function with a name like `_switch`:

```cpp
switch(my_enum._switch()) {
    case Foo::A: break;
    case Foo::B: break;
}
```

We could also make this `operator +`, if you can bear all the `operator +`s :)

```cpp
switch(+my_enum) { ...
```

The conversion this would perform would be to an unscoped enum type, so we could use the `Foo::A` syntax in cases without having to prepend the existing `operator +`.

cc @cheparukhin @mbevin @rcdailey

**Comment by aantron** [comment]
Re: @rcdailey's suggestion of enumerators from #23:

I considered enumerators, but IIRC the preprocessor+template hack I use to get mixed initialization syntax (`{A, B, C = 4, D}`) working doesn't work very well with them, and I had convinced myself that there wasn't a variant that does. I may be wrong, as it has been a while, but for background on that part of Better Enums, you can see this [answer](http://stackoverflow.com/a/31362042) on SO.

---

Re: @rcdailey on breaking compatibility with C++03/98:

I should repost from #22:

> The guideline I am using is that the enum.h in master should be immediately compatible with recent versions of GCC, Clang, and MSVC. [snip]
>
> Since MSVC has had enum class for a while now, we can switch to that, and drop support for [converting to] the old-style internal enum completely. After doing that, we can just add a note to the README telling users of older MSVC versions which release of Better Enums they should download.

The same guideline would probably preclude switching to C++11 completely (i.e. `constexpr`), due to the infamous #9.

> If you want to know the status of that, do an amend or add a dummy commit in the vc2015-constexpr branch to trigger a build, and see what happens in AppVeyor. You might have to look up the most recent version of MSVC, and check that the AppVeyor image we are using has that.

So it's not so much the age of C++11 that constrained me, as how long it has been since MSVC added support for each particular feature. Have you heard anything about MS potentially switching to Clang recently?

Also, the approach I suggested when opening this issue won't cause Better Enums to stop supporting C++03/98, though we might drop it anyway on another count. But either approach is fine.

---

Concerning the internal type:

I think it actually doesn't matter *that much* what that is, whether enumerator, scoped enum, or unscoped enum, as long as a Better Enum can be introduced naturally (`Foo::A`) and used naturally, or close to that, in `switch`. The latter seems to imply that there must be some kind of conversion to some kind of enum type, as otherwise you lose exhaustiveness checking. The current interface is, as we all agree, really leaky, since it has an implicit conversion to an unscoped enum. AFAIK it is the only way to get fully natural `switch` syntax, but I think a lot of people are willing to forgo that. What are your opinions?

**Comment by cheparukhin** [comment]
I don't like the discrepancy between external and internal enum types because of all these conversion issues, so I'd rather try to have just a single type.

Some crude ideas from the top of my head:

1. Use `enum class` for enum type and augment it with free functions in the macro:

    ```
    enum class MyEnum: uint8_t { ONE, TWO, THREE }; 
    MyEnum toMyEnum(const std::string&);
    std::string toString(MyEnum);
    ```

    This will obviously break API and it means that almost the whole library should be rewritten, so I guess it's not viable.

2. Define enum values as static constants inside the enum type, using the same type:

    ```
    class MyEnum {
    public:
        static const MyEnum ONE;
        static const MyEnum TWO;
    
    private:
        constexpr MyEnum(uint8_t value): m_value(value) {}
        uint8_t m_value;
    };

    const MyEnum MyEnum::ONE = 1;
    const MyEnum MyEnum::TWO = 2;
    ```

    What bothers me is that it's not possible to use `constexpr` in this case as `MyEnum` would be incomplete at the point of these definitions.

@aantron How do you reckon?

**Comment by aantron** [comment]
> Use enum class for enum type and augment it with free functions in the macro:

Let me paraphrase to make sure I understood you correctly: you want the macro to generate a normal, non-wrapped `enum`, and then also generate free functions that do the various conversions and whatnot that Better Enums provides. If so, this is a good idea, and it would eliminate a ton of problems.

I was calling this "enum traits," because in this mode, Better Enums is generating something like a traits `struct` over a regular `enum`. This is opposed to the current approach, where a `struct` wraps an `enum`, and you use objects of this `struct` type.

There is an old implementation of traits in a [branch](https://github.com/aantron/better-enums/tree/traits). I decided not to pursue it at the time for [various reasons](http://aantron.github.io/better-enums/DesignDecisionsFAQ.html#ShouldBetterEnumsProvideEnum). It was discussed multiple times on Reddit and in email. I later decided that it's worth maintaining side-by-side with the existing "object" `enum.h` (see #17). But, then I got burned out, so that didn't go anywhere.

I wouldn't worry about breaking the API. It can be implemented in a branch, or in a separate file `enum_traits.h`, or whatever. The separate file is probably the better. We can let people know, and if people definitively prefer traits, Better Enums can drop the object approach. It is extra maintenance work, but the experiment is probably worth it.

---

> Define enum values as static member constants of the same type:

I'm not sure this can work. Consider some of the constraints:

1. One of the strengths of Better Enums is that it uses the already-familiar (and pretty decent) `enum` declaration syntax, i.e. `A, B, C = 4, D`.

2. Each of `A`, `B`, `C = 4`, `D` becomes a macro paramter.

3. There is no way, AFAIK, to separate the name `C` in the macro parameter from the `= 4` initializer. This already seems to preclude

        static const MyEnum TWO;

  since you have to paste in both name and initializer, if the latter is present. A further problem is you don't really know what values to assign to `ONE`, `TWO`. Using my example, how would you know that `D` should be `5`?

I've considered fancier schemes with enumerators, like defining an enum and then referring to it when declaring enumerators:

        enum Internal {A, B, C = 4, D};
        // ...
        static const Whatever A = Internal::A
        static const Whatever B = Internal::B
        static const Whatever C = 4 = Internal::C = 4   // wat

As you can see, that breaks down. In the last line, casting to dummy types with overridden `operator =` can eliminate all the problems, except, AFAIK, for the problem of assigning to `4` (because, AFAIK, you cannot get a cast on either the `C` or the `4`, due to this being declaration syntax, and `C = 4` coming from a macro argument; you cannot have free overloads of `operator =`).

There are several other variations with enumerators, but all that I considered seem to suffer from variations of this problem.

---

If you want to work on the "object"-style enums, the best advice I can give is to start with the code in the SO [answer](http://stackoverflow.com/a/31362042) I linked. It's pretty short (&lt;200 lines including whitespace), and includes only the core of Better Enums, with only two functions implemented on top of that. You can use it as a small example to test modifications, or as a reference for writing an approach from scratch. It should quickly filter out approaches that don't pass the (weird) constraints of C++.

**Comment by rcdailey** [comment]
After reading more into the discussion, I agree with @cheparukhin's first idea. I think it would make sense to keep the scoped enum outside and just build a series of free functions to operate on them. This is especially easy to do with operators such as `operator<<` and `operator>>`, which are arguably all you really need. STL and Boost use external stream operators on types for conversion to/from string format. So you could easily do:

```
BETTER_ENUM(TestEnum, int, One, Two, Three);

std::string one_str = boost::lexical_cast<std::string>(TestEnum::One);
```

Naturally, switch will already work *and* you get type-safety guarantees.

Are there any current requirements that would be difficult or impossible to support using this approach? Honestly the only reason I really use better enums in production code is for the convenient string to enum mapping.

I think this functionality will be built into the language by C++17 or C++20, though.

**Comment by aantron** [comment]
Indeed, that's what is in the `traits` branch now. I just didn't keep that branch up-to-date with subsequent improvements to the internals of `master`.

See:

- [basic usage](https://github.com/aantron/better-enums/blob/traits/samples/1-basic.cc)
- [switch](https://github.com/aantron/better-enums/blob/traits/samples/3-switch.cc)
- [all the samples](https://github.com/aantron/better-enums/tree/traits/samples)

> Are there any current requirements that would be difficult or impossible to support using this approach? 

Besides what's in the linked [FAQ](http://aantron.github.io/better-enums/DesignDecisionsFAQ.html#ShouldBetterEnumsProvideEnum), I think not. You lose:

- reflective enums in namespaces other than the global namespace, unless we can come up with a workaround,
- a small degree of conciseness in the syntax,
- the ability to prevent default construction, for people who care about that.

This is properly issue #17, but basically, getting the `traits` branch updated would involve going through the history since May 17 or so, and applying anything that makes sense. For example:

- Inlining the map macro (analog of 460bb806a38f9a583882fdbb4484b3ce8a954160), for a single-header library.
- Renaming the main macro (#11).
- Various internal improvements, such as removing weak symbols.

**Comment by aantron** [comment]
Regarding C++17, it will greatly simplify the internals of the library. I am not sure what, at this point, exactly is planned for C++17, or whether it will have to wait until C++20. But, the last I checked, the interfaces proposed were pretty basic and difficult to use directly, so it may still be worthwhile to have Better Enums around. Otherwise, one of those standards will retire this library :)
